company-lsp.el

```diff
@@ -139,6 +143,32 @@ Returns CANDIDATE with the resolved CompletionItem."
                              candidate)))))
   candidate)
 
+(defun company-lsp--rust-completion-snippet (item)
+  "Function providing snippet with the rust language.
+It parses the function's signature in ITEM (a CompletionItem)
+to expand its arguments."
+  (let* ((kind (gethash "kind" item))
+         (is-function (when (numberp kind) (= kind 3)))
+         (detail (gethash "detail" item)))
+    (when is-function
+      (let ((snippet (when (and detail (s-matches? "^\\(pub \\)?\\(unsafe \\)?fn " detail))
+                       (-some--> (substring detail (1+ (s-index-of "(" detail)) (s-index-of ")" detail))
+                                 (replace-regexp-in-string "^[^,]*self\\(, \\)?" "" it)
+                                 (s-split ", " it)
+                                 (mapconcat (lambda (x) (format "${%s}" x)) it ", ")))))
+        (concat "(" (or snippet "$1") ")$0")))))
```

Comment:
Do you need $1 if snippet is nil? It's better to just stop at the end of right parenthesis for functions without parameters.

Reply:
I was thinking that if, for some reason, the RLS doesn't give us the signature, it would still be better to stop the cursor between the parenthesis..